src/compiler/checker.ts

```diff
@@ -4825,6 +4763,88 @@ namespace ts {
             return widened;
         }
 
+        function getJSDocTypeFromSpecialDeclarations(declaredType: Type | undefined, expression: Expression, symbol: Symbol, declaration: Declaration) {
+            const typeNode = getJSDocType(expression.parent);
+            if (typeNode) {
+                const type = getWidenedType(getTypeFromTypeNode(typeNode));
+                if (!declaredType) {
+                    return type;
+                }
+                else if (declaredType !== errorType && type !== errorType &&
+                         !isTypeIdenticalTo(declaredType, type) &&
+                         !(symbol.flags & SymbolFlags.JSContainer)) {

```

Comment:
I don't understand this contdition? Why do we ignore the type when it's in a `JSContainer`?

Reply:
That's ... uh ... a good question. Removing it doesn't cause any tests to fail. I added it in #20198, but I think there's enough test coverage in that PR that removal is safe.